src/main/java/htsjdk/variant/variantcontext/writer/BCF2Writer.java

```diff
@@ -221,6 +206,43 @@ public void close() {
         super.close();
     }
 
+    @Override
+    public void setVcfHeader(VCFHeader header) {
+        if (isHeaderWritten) {
+            throw new IllegalStateException("Header is already written");
+        }
+        if (isBodyWritten) {
+            throw new IllegalStateException("Body is already written");
+        }
```

Comment:
As mentioned above, I think we need only one state variable. If we can get away with that, this message can say something like "The header cannot be modified after the header or variants have been written to the output stream."